adserver/management/commands/archive_offers.py

```diff
@@ -0,0 +1,229 @@
+from django.utils import timezone
+from django.utils.translation import ugettext_lazy as _
+
+
+class Command(BaseCommand):
+
+    """Management command to help archive offers."""
+
+    help = "Archives offers in CSV form from the database to files: one per day."
+
+    default_output_dir = "/tmp"
+    default_end_date = (timezone.now() - datetime.timedelta(days=90)).date()
+    default_start_date = default_end_date - datetime.timedelta(days=30)
+
+    storage_output_dir = "offers/"
+
+    # This will be populated from the default or command line args
+    output_dir = None
+
+    def _from_isoformat(self, date_str):
+        """
+        Returns a datetime.date objects for a given date string.
+
+        This method was added to datetime.date in Python 3.7. Use that after upgrading.
+        """
+        return datetime.datetime.strptime(date_str, "%Y-%m-%d").date()
+
+    def add_arguments(self, parser):
+        """Add command line args for this command."""
+        parser.add_argument(
+            "-o",
+            "--output-dir",
+            default=self.default_output_dir,
+            type=str,
+            help=_("Directory to write output files (a subdir will be created)"),
+        )
+        parser.add_argument(
+            "-d",
+            "--delete-offers",
+            action="store_true",
+            default=False,
+            help=_(
+                "Delete offers after archiving (requires successful copying to remote backups)"
+            ),
+        )
+        parser.add_argument(
+            "-s",
+            "--start-date",
+            default=self.default_start_date,
+            type=self._from_isoformat,
+            help=_("Start date to dump offers (default 7 days ago)"),
+        )
+        parser.add_argument(
+            "-e",
+            "--end-date",
+            default=self.default_end_date,
+            type=self._from_isoformat,
+            help=_("End date to dump offers (inclusive, defaults to yesterday)"),
+        )
+
+    def handle_archive_day(self, day):
+        """Archive a single day of offers to a file."""
+        output_file = self.output_dir / f"{day:%Y-%m-%d}-offers.csv"
+        gzip_output_file = Path(str(output_file) + ".gz")
+        end_day = day + datetime.timedelta(days=1)
+
+        self.stdout.write(_("Archiving %s to %s...") % (day, output_file))
+
+        # Using the date params as an f-string is suboptimal but these are validated
+        query = f"""
+            COPY (
+                SELECT * FROM adserver_offer
+                WHERE date >= '{day:%Y-%m-%d}' AND date < '{end_day:%Y-%m-%d}'
+                ORDER BY date
+            ) TO STDOUT WITH CSV HEADER"""
+        with connection.cursor() as cursor:
+            with open(output_file, "w") as fd:
+                # https://www.psycopg.org/docs/cursor.html#cursor.copy_expert
+                cursor.copy_expert(query, fd)
+
+        # This will be off by one because the CSV contains a header row
```

Comment:
Off by 1 compared to what? The number output by the SQL command?

Reply:
`wc -l` is outputting the lines in the file which will be one more than the actual number of records. The SQL copy doesn't output a number.